### DISCUSSION TITLE
Confusion about Permission Registration + Gate

### DISCUSSION THREAD

**Discussion by connormcmanustw** [spatie/laravel-permission#2640] [Q&A]
I have recently shifted from an old Laravel 8 application up to Laravel 11, as well as updated this package to v6. 

Upon inspection of what the old app was doing, after deep-diving into the base Laravel `Gate` class `(Illuminate\Auth\Access\Gate.php)`, the `PermissionRegistrar` from this package was being included in the "beforeCallbacks" array of that GateClass. See below:

<img width="850" alt="Screenshot 2024-03-18 at 8 20 08 PM" src="https://github.com/spatie/laravel-permission/assets/36934584/bf0482c8-6bcf-40ae-8d7f-ffcaaac1830a">

This is the dump output:
<img width="571" alt="Screenshot 2024-03-18 at 8 21 11 PM" src="https://github.com/spatie/laravel-permission/assets/36934584/a4a53cb0-1de7-4eee-8926-b0e81d7c2384">

However, after shifting + updating the package, that dump is empty, and the `PermissionRegistrar` is no longer included. This is causing a different flow in the conditional, as seen in the snippet above:
```
$result = $this->callBeforeCallbacks(
    $user, $ability, $arguments
);

 if (is_null($result)) {
    $result = $this->callAuthCallback($user, $ability, $arguments);
}
```
... `$result` is `null`, since there are no `beforeCallbacks` array items (where the `PermissionRegistrar` previously was), and causing the if statement to hit, which, just based on what I'm seeing, is the issue. 

This difference is causing a lot of my blade directives to fail, and I'm just trying to figure out what exactly I need to change.

**Comment by drbyte** [comment]
In v6, in `PermissionRegistrar`, we still call `$gate->before()`, where `$gate` is lazy-bound so it's resolved when required, not prematurely. Ref #2321 
In prior versions we forced immediate resolving by using `app(Gate::class)->before()`.
